pkg/scheduler/framework/plugins/volumezone/volume_zone.go

```diff
@@ -243,50 +344,103 @@ func (pl *VolumeZone) Filter(ctx context.Context, cs *framework.CycleState, pod
 		}
 		if !ok || !pvTopology.values.Has(v) {
 			logger.V(10).Info("Won't schedule pod onto node due to volume (mismatch on label key)", "pod", klog.KObj(pod), "node", klog.KObj(node), "PV", klog.KRef("", pvTopology.pvName), "PVLabelKey", pvTopology.key)
-			return framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrReasonConflict)
+			return framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrReasonConflict), false
 		}
 	}
-
-	return nil
+	return nil, true
 }
 
-func getStateData(cs *framework.CycleState) (*stateData, error) {
-	state, err := cs.Read(preFilterStateKey)
+// isSchedulableAfterPersistentVolumeClaimAdded is invoked whenever a PersistentVolumeClaim added. It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (pl *VolumeZone) isSchedulableAfterPersistentVolumeClaimAdded(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	_, addedPVC, err := util.As[*v1.PersistentVolumeClaim](nil, newObj)
 	if err != nil {
-		return nil, err
+		logger.Error(err, "unexpected objects in isSchedulableAfterPersistentVolumeClaimAdded", "oldObj", oldObj, "newObj", newObj)
+		return framework.QueueAfterBackoff
 	}
-	s, ok := state.(*stateData)
-	if !ok {
-		return nil, errors.New("unable to convert state into stateData")
+
+	pvcNames := pl.getPersistentVolumeClaimClassNameFromPod(logger, pod)
+	for _, pvcName := range pvcNames {
+		if (addedPVC.Name == pvcName) && (pod.Namespace == addedPVC.Namespace) {
+			logger.V(4).Info("PVC was created, and matches the pod's PVC", "pod", klog.KObj(pod), "PVC", klog.KObj(addedPVC))
+			return framework.QueueAfterBackoff
+		}
 	}
-	return s, nil
+
+	logger.V(4).Info("PVC was created, but it doesn't match the pod's PVC", "pod", klog.KObj(pod), "PVC", klog.KObj(addedPVC))
+	return framework.QueueSkip
 }
 
-func getErrorAsStatus(err error) *framework.Status {
+// isSchedulableAfterPersistentVolumeChange is invoked whenever a PersistentVolume added or updated. It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (pl *VolumeZone) isSchedulableAfterPersistentVolumeChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	originalPV, modifiedPV, err := util.As[*v1.PersistentVolume](oldObj, newObj)
 	if err != nil {
-		if apierrors.IsNotFound(err) {
-			return framework.NewStatus(framework.UnschedulableAndUnresolvable, err.Error())
+		logger.Error(err, "unexpected objects in isSchedulableAfterPersistentVolumeChange", "oldObj", oldObj, "newObj", newObj)
+		return framework.QueueAfterBackoff
+	}
+	originalPVTopologies := make([]pvTopology, 0)
+	if originalPV != nil {
+		originalPVTopologies, _ = pl.getPVTopologiesAndSort(logger, pod, originalPV)
+	}
+	modifiedPVTopologies, isPodBoundToPV := pl.getPVTopologiesAndSort(logger, pod, modifiedPV)
+	if isPodBoundToPV && !reflect.DeepEqual(originalPVTopologies, modifiedPVTopologies) {
```

Comment:
I suggest to write a new function to check if two PVTopologies are equal. Reflect is convenient yes, but it will cost more.

Reply:
@AxeZhan ,
Just to clarify, what is your suggestion?
Should I create the function `CheckTopologiesEqual`, and within this function, call both the `sort` function and `reflect.DeepEqual`?
Or do you mean I should not use `reflect.DeepEqual` and create a similar function from scratch?
It would be great if there are more useful library than `reflect.DeepEqual`. However, I don't think it is convenient to create function similar to `reflect.DeepEqual` from scratch.

Reply:
I think we should try to avoid `reflect here`. 
We only need to check pvTopolgies here, and we don't need to create a common function.
For that, we can simply compare the pvName,key and values of a topology, since you've sorted them before.
And set itself provides a `equal` function. So I think it should be fine.
```
// Equal returns true if and only if s1 is equal (as a set) to s2.
// Two sets are equal if their membership is identical.
// (In practice, this means same elements, order doesn't matter)
func (s1 Set[T]) Equal(s2 Set[T]) bool {
	return len(s1) == len(s2) && s1.IsSuperset(s2)
}
```

Reply:
@AxeZhan ,
Thanks for your explanation. I have an additional question.

Question:
Why should we try to avoid `reflect.DeepEqual` ? Is there any potential issues in `reflect`?
I found the library about Set and IsSuperset [here](https://github.com/ncw/gotemplate/blob/be291eb507fd/set/set.go#L192). Looks like this library is ncw's personal project. I thought that the `reflect` package would be more reliable than Set and IsSuperset because `reflect` package is included in Go itself.

Reply:
https://github.com/kubernetes/kubernetes/pull/119373#discussion_r1268371229
BTW, I found we had previously discussed here

Reply:
> Why should we try to avoid reflect.DeepEqual ? Is there any potential issues in reflect?

My concern is that there is a performance cost to using reflect. Your point makes some sense too.

> I found the library about Set and IsSuperset 

Library about Set and IsSuperset is vendered `"k8s.io/apimachinery/pkg/util/sets"`, and we used it in many places already, I think it's also reliable.

>  BTW, I found we had previously discussed here

That's an another discussion, it's to suggest that we can't using reflect.deepequal directly(because we have to ignore the order of topology). And is solved by adding a sort before :)

Anyway, it's just my concern.This doesn't affect the correctness of this function. Let's leave it for the following discussion.

Reply:
Thank you so much for your explanation! I got what you meant.

> Library about Set and IsSuperset is vendered "k8s.io/apimachinery/pkg/util/sets", and we used it in many places already, I think it's also reliable.

I didn't know that, thank you! While waiting for another review, I will investigate on this library :)